pages/blog/socketio-part1.md

```diff
@@ -0,0 +1,64 @@
+---
+type: Engineering
+title: "The journey of documenting a Socket.IO API (Pt 1)"
+date: 2021-09-22T06:00:00+01:00
+featured: true
+cover: /img/posts/socketio-part1/cover.webp
+tags:
+   - Specification
+   - Socket.IO
+   - Protocol
+   - Asynction
+authors:
+   - name: Dimitrios Dedoussis
+     photo: /img/avatars/dedoussis.webp
+     link: https://twitter.com/dedoussis
+     byline: Senior Software Engineer at Babylon & Maintainer of Asynction
+---
+
+> This post originally appeared on [https://dedouss.is](https://dedouss.is/posts/2021-06-24-documenting-socketio-part-1.html)
+
+My recent adventures with [Socket.IO](https://socket.io/) took me off on a tangent of exploring the world of [AsyncAPI](https://www.asyncapi.com/) as a means of documenting Socket.IO APIs. This is the first part of a series of blog posts covering the modelling of the Socket.IO protocol using AsyncAPI objects followed by a step-by-step tutorial on how to create a specification YAML file given an existing Socket.IO API.
+
+## Setting the scene
+
+Earlier this year my team undertook the task of re-implementing the backend of a realtime chat application – mentioning _re-implementing_ and _backend_ in the very first sentence is probably a PTSD trigger for many of you, but thankfully this post has nothing to do with rewriting-from-scratch horror stories. The app was originally built using a 3rd party push-notifications platform which allowed us to deliver a functional MVP in a relatively speedy manner. However, as new requirements started creeping in from the business, it was clear to the team that this 3rd party dependency was not really worth it anymore. We thus had to take the (arguably not so easy) decision to implement our own realtime API to gain complete control of each server connection. The design meeting would soon follow, to answer questions like “What framework should we use?” or “Would serverless make sense?”. The stack of the team, namely Python, [Flask](https://flask.palletsprojects.com/) and [Kubernetes](https://kubernetes.io/), pointed us in the direction of [Flask-SocketIO](https://flask-socketio.readthedocs.io/), making Socket.IO the protocol of choice. Although pragmatic, this stack driven design approach felt very unorthodox. The tooling dictated the choice of the client<->server communication protocol rather than the other way around. It was a decision primarily based on the stack and the expertise of the backend team. What if the Socket.IO client library for Swift is not maintained anymore? What if Socket.IO is a very verbose protocol, not friendly for clients with limited network connectivity? Thankfully, this is not the case and Socket.IO happens to be a well supported and carefully designed protocol. Nonetheless, the decision was made and the development of a Socket.IO API was about to commence.
```

Comment:
```suggestion
Earlier this year my team undertook the task of re-implementing the backend of a realtime chat application – mentioning _re-implementing_ and _backend_ in the very first sentence is probably a PTSD trigger for many of you, but thankfully this post has nothing to do with rewriting-from-scratch horror stories. The app was originally built using a 3rd party push-notifications platform which allowed us to deliver a functional MVP in a relatively speedy manner. However, as new requirements started creeping in from the business, it was clear to the team that this 3rd party dependency was not really worth it anymore. We thus had to take the (arguably not so easy) decision to implement our own realtime API to gain complete control of each server connection. The design meeting would soon follow, to answer questions like _“What framework should we use?”_ or _“Would serverless make sense?”_. The stack of the team, namely Python, [Flask](https://flask.palletsprojects.com/) and [Kubernetes](https://kubernetes.io/), pointed us in the direction of [Flask-SocketIO](https://flask-socketio.readthedocs.io/), making Socket.IO the protocol of choice. Although pragmatic, this stack-driven design approach felt very unorthodox. The tooling dictated the choice of the client<->server communication protocol rather than the other way around. It was a decision primarily based on the stack and the expertise of the backend team. What if the Socket.IO client library for Swift is not maintained anymore? What if Socket.IO is a very verbose protocol, not friendly for clients with limited network connectivity? Thankfully, this is not the case and Socket.IO happens to be a well supported and carefully designed protocol. Nonetheless, the decision was made and the development of a Socket.IO API was about to commence.
```